Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,219 @@
+                }
+                let newCoordinate = currentCoordinate.nextRow
+                tetrisGrid.addTile(at: newCoordinate, tileText: text)
+                coordinatesToShift.append((from: currentCoordinate, to: newCoordinate))
+            }
+        }
+        delegate?.animate(shiftTiles: coordinatesToShift)
+    }
+
+    // TODO: fetch from database to match valid phrase
+    private func getValidPhraseRange(_ line: String) -> CountableRange<Int>? {
+        let phrases = ["先发制人"]
+        for phrase in phrases {
+            let phraseLen = phrase.characters.count
+            for i in 0...line.characters.count - phraseLen {
+                let startIndex = line.index(line.startIndex, offsetBy: i)
+                if line[startIndex..<line.index(startIndex, offsetBy: phraseLen)] == phrase {
+                    return i..<i + phraseLen
+                }
+            }
+        }
+        return nil
+    }
+
+    // TODO: generate from database, base on existing grid
+    fileprivate func getRandomWord() -> String {
+        let words = "先发制人"
+        return String(words[words.index(words.startIndex,
+                                        offsetBy: Random.integer(toInclusive: words.characters.count))])
+    }
+
+    @inline(__always)
+    fileprivate func assertCoordinateValid(_ coordinate: Coordinate) {
+        assert(coordinate.row < Constants.Game.Tetris.rows &&
+               coordinate.col < Constants.Game.Tetris.columns, "Coordinate is out of bounds")
+    }
+}
+
+extension TetrisGameViewModel: TetrisGameViewModelProtocol {
+
+    var gameTitle: String {
+        return Constants.Game.Tetris.gameTitle
+    }
+
+    var gameInstruction: String {
+    	return Constants.Game.Tetris.gameInstruction
+    }
+
+    func canShiftFallingTile(to coordinate: Coordinate) -> Bool {
+        assertCoordinateValid(coordinate)
+        return !tetrisGrid.hasTile(at: coordinate)
+    }
+
+    func dropNextTile() -> (location: Coordinate, tileText: String) {
+        upcomingTiles.append(getRandomWord())
+        let tileText = upcomingTiles.removeFirst()
+        delegate?.redisplayUpcomingTiles()
+
+        fallingTileText = tileText
+        let randCol = Random.integer(toInclusive: Constants.Game.Tetris.columns)
+        return (location: Coordinate(row: Coordinate.origin.row, col: randCol),
+                tileText: tileText)
+    }
+
+    func canLandTile(at coordinate: Coordinate) -> Bool {
+        assertCoordinateValid(coordinate)
+        if tetrisGrid.hasTile(at: coordinate) ||
+           (!tetrisGrid.hasTile(at: coordinate.nextRow) && coordinate.row < Constants.Game.Tetris.rows - 1) {
+            return false
+        }
+        landingCoordinate = coordinate
+        return true
+    }
+
+    func tileHasLanded() {
+        guard let fallingTileText = fallingTileText,
+              let landingCoordinate = landingCoordinate else {
+            fatalError("fallingTileText or landingCoordinate is not initialised")
```

Comment:
alignment issue: `fatalError` is of an alternate code path than `let landingCoordinate = landingCoordinate`, but indented earlier than `let landingCoordinate = landingCoordinate`